src/libs/elektra/kdb.c

```diff
@@ -1006,6 +1007,14 @@ KDB * kdbOpen (const KeySet * contract, Key * errorKey)
 		goto error;
 	}
 
+	// TOOD (atmaxinger): improve
+	if (initHooks (handle, ksDup (elektraKs), handle->modules, errorKey) == -1)
```

Comment:
`initHooks` only has a single `return 0`, this `if` is always false

Reply:
For now ;)

Reply:
it will now return -1 if any hooks fail to initialized

Reply:
`kdb cache` already has support for disabling a global plugin, this feature should be available (in the same way) for all global plugins.

Important is that the user can always fix the problem somehow (e.g. disable the broken plugin). If we already fail in kdbOpen, then e.g. `kdb cache disable` probably will also fail, as it tries to write to KDB (and obviously needs to open it first).

So I think broken global plugins can only lead to a warning, not to an error.

Reply:
Yes, broken plugins only lead to warnings, this is implemented right now.

> kdb cache already has support for disabling a global plugin, this feature should be available (in the same way) for all global plugins.

We do. As I see, `kdb cache enable/disable` sets a key in the contract. This contract is also passed to `initHooks`. We already parse the contract to enable/disable the `gopts` plugin, so this is accounted for in the current architecture.

Reply:
> So I think broken global plugins can only lead to a warning, not to an error.

They already do.

> kdb cache already has support for disabling a global plugin, this feature should be available (in the same way) for all global plugins.

Agreed. The way `kdb cache` does it, it sets the key `system:/elektra/cache/enabled` to either `0` or `1.`.  We can just pass the config to `initHooks` and then check there for the respective keys.

One question that comes to my mind: What shall we do if those keys are not present for a plugin? Is it enabled or disabled by default?

Reply:
See doc/decisions/boolean.md a boolean is by default 0 but we can set all hooks to be 1 via a installation script, see #4446.

Btw. we should create a sub-hierarchy system:/elektra/hook/cache/enabled (if we settle to the global plugins called hooks, is this now fixed?)

Reply:
> Btw. we should create a sub-hierarchy system:/elektra/hook/cache/enabled (if we settle to the global plugins called hooks, is this now fixed?)

Sure. But do we want `hook` (singular) or `hooks` (plural) for the sub-hierarchy? The same would also apply for the `system:/elektra/modules/<plugin name>/exports/hooks/<hook name>/get`. Currently I use the plural form there to be consistent with `modules` and `exports`.

Reply:
> it will now return -1 if any hooks fail to initialized

What I meant was that there is already a failure case (couldn't load `gopts` plugin), but the function doesn't actually return an error code for that.

> this feature should be available (in the same way) for all global plugins.

I agree and disagree... The flag in the KDB doesn't make sense for all hooks. Especially for `gopts` it isn't the right solution. `gopts` can only be enabled via the `kdbOpen` contract. Therefore having a permanent flag in the KDB makes zero sense. It makes no sense to "enable gopts by default" and if the flag is used to override the contract it will just break applications that expect `gopts` to run.

For `cache` the flag makes sense of course. For `spec` I could see the flag as well, but in this case it is probably more about switching from errors to warnings. I don't really see a case where completely disabling `spec` would help. Finally, there is the notification stuff. AFAIK for `internalnotification` the situation is the same was with `gopts`. You enable it only via contract and therefore a flag in the KDB is useless. For `dbus` and `zeromq` enabling/disabling via the KDB would probably make sense again.

> system:/elektra/hook/cache/enabled

Yes, I'd do that too. But as stated above, I'd tailor the actual config to the individual hook. So for `spec` it might be something like `system:/elektra/hook/spec/loglevel`.

> we can set all hooks to be 1 via a installation script,

We could also switch from `cache/enabled` to `cache/disabled` which automatically switches the default to "cache enabled".

Reply:
> do we want `hook` (singular) or `hooks` (plural) for the sub-hierarchy?

`doc/DESIGN.md` says we use singular for all names

https://github.com/ElektraInitiative/libelektra/blob/ce6c6984a81037559332640258c7bd3774ce1036/doc/DESIGN.md?plain=1#L194

But it is not done very consistently, especially for keynames. I also think that often the plural feels more natural in keynames (since it refers to a collection of things), but always using singular is probably the easiest rule to have.

Reply:
Yes, let us stick to always singular. If you spot somewhere plural, we can fix it before 1.0. I created  #4472 to collect such issues.

In some cases it feels a bit weird (e.g. /etc/host) but I think it is only a matter of getting used to it.

Reply:
About enabling plugins: if it does not make sense to enable a plugin, we don't need to provide a flag (and the plugin will always be disabled except if enabled via contract). gopts is the only one, isn't it?

Reply:
> gopts is the only one, isn't it?

As stated above, I think `internalnotification` also needs to be enabled via contract, because it needs a function pointer to the callback(s).

Reply:
> Yes, let us stick to always singular. 

Ok, I have updated the newly added key in the exports.